File: lib/features/activity_packrafting/models/packrafting_conditions_report.dart
import 'package:turbo/features/activity_fishing/api.dart' show WeatherSlice;

class PackraftingConditionsReport {
  final String activityId;
  final DateTime validAt;
  final DateTime fetchedAt;
  final WeatherSlice weather;

  /// Current flow in m³/s from the linked NVE station. Null when the
  /// activity has no station configured or the NVE provider isn't wired.
  final double? currentFlowCumecs;

  /// "rising" / "stable" / "falling", or null if not enough data.
  final String? flowTrend;

  final int? score;
  final String rationale;

  const PackraftingConditionsReport({
    required this.activityId,
    required this.validAt,
    required this.fetchedAt,
    required this.weather,
    required this.currentFlowCumecs,
    required this.flowTrend,
    required this.score,
    required this.rationale,
  });

  factory PackraftingConditionsReport.fromJson(Map<String, dynamic> json) =>
      PackraftingConditionsReport(
        activityId: json['activityId'] as String,
        validAt: DateTime.parse(json['validAt'] as String),
        fetchedAt: DateTime.parse(json['fetchedAt'] as String),
        weather: WeatherSlice.fromJson(json['weather'] as Map<String, dynamic>),
        currentFlowCumecs: (json['currentFlowCumecs'] as num?)?.toDouble(),
        flowTrend: json['flowTrend'] as String?,
        score: (json['score'] as num?)?.toInt(),
        rationale: json['rationale'] as String,
      );
}

File: src/Activities/Turbo.Activities.Packrafting.Contracts/value/PackraftingConditionsReport.cs
using System.Text.Json.Serialization;
using Turboapi.Activities.value;

namespace Turboapi.Activities.Packrafting.value;

public sealed record PackraftingConditionsReport
{
    [JsonPropertyName("activityId")] public Guid ActivityId { get; init; }
    [JsonPropertyName("validAt")] public DateTimeOffset ValidAt { get; init; }
    [JsonPropertyName("fetchedAt")] public DateTimeOffset FetchedAt { get; init; }
    [JsonPropertyName("weather")] public WeatherSlice Weather { get; init; }

    /// <summary>Current flow in m³/s from the linked NVE station. Null
    /// when no station is configured or the NveRiverFlowProvider isn't
    /// running.</summary>
    [JsonPropertyName("currentFlowCumecs")] public float? CurrentFlowCumecs { get; init; }

    /// <summary>Trend over the last 24h: "rising", "stable", "falling"
    /// or null if not enough data.</summary>
    [JsonPropertyName("flowTrend")] public string? FlowTrend { get; init; }

    [JsonPropertyName("score")] public int? Score { get; init; }
    [JsonPropertyName("rationale")] public string Rationale { get; init; }

    [JsonConstructor]
    public PackraftingConditionsReport(
        Guid activityId, DateTimeOffset validAt, DateTimeOffset fetchedAt,
        WeatherSlice weather, float? currentFlowCumecs, string? flowTrend,
        int? score, string rationale)
    {
        ActivityId = activityId;
        ValidAt = validAt;
        FetchedAt = fetchedAt;
        Weather = weather;
        CurrentFlowCumecs = currentFlowCumecs;
        FlowTrend = flowTrend;
        Score = score;
        Rationale = rationale ?? throw new ArgumentNullException(nameof(rationale));
    }
}
